File: app/channels/stimul8/component_channel.rb
module Stimul8
  class ComponentChannel < ApplicationCable::Channel
    attr_reader :component

    def subscribed
      @component = Stimul8::Component.recreate params[:component_class], params[:component_id], context: context
    end

    def unsubscribed
      @component = nil
    end

    def call_method data
      data = data.symbolize_keys
      data.delete :action
      method_name = data.delete :method_name
      @component.call_method(method_name, **data)
    end
  end
end

File: app/javascript/channels/stimul8/component_channel.js
import consumer from "channels/consumer"
import { Turbo } from "@hotwired/turbo-rails"

export default componentSubscription = (componentId, componentClass, element) => consumer.subscriptions.create(
  { channel: "ComponentChannel", componentClass: componentClass, componentId: componentId },
  {
    connected() {
      // Called when the subscription is ready for use on the server
    },
    disconnected() {
      // Called when the subscription has been terminated by the server
    },
    received(data) {
      if (data.type == "redraw") {
        element.outerHTML = data.html
      } else if (data.type == "remove") {
        element.remove()
      } else if (data.type == "reload") {
        Turbo.visit(window.location.href, { action: "replace" })
      } else if (data.type == "redirect") {
        Turbo.visit(data.redirectUrl)
      }
      // TODO: add in types for animations and events - and add event handlers to components
    },
    callMethod(params) {
      this.perform("call_method", params)
    }
  }
);
